File: calendar-conflicts.js
(function (root, factory) {
  const api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  if (root) root.CalendarConflicts = api;
})(typeof window !== 'undefined' ? window : globalThis, function () {
  const DERIVED_SOURCES = new Set(['ai', 'deadline', 'email', 'attention', 'work', 'virtual']);

  function parseDate(value) {
    if (!value || String(value).length <= 10) return null;
    const date = new Date(value);
    return Number.isNaN(date.getTime()) ? null : date;
  }

  function interval(event) {
    const start = parseDate(event?.start);
    const end = parseDate(event?.end);
    return start && end && start < end ? { start, end } : null;
  }

  function marker(event) {
    const harness = event?.agent_harness || {};
    return String(event?.marker || event?.source_message_id || harness.agent_harness_marker || harness.source_message_id || '').trim();
  }

  function normalizedTitle(event) {
    return String(event?.title || '').toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim();
  }

  function isBlocking(event) {
    const source = String(event?.source || 'google').toLowerCase();
    if (DERIVED_SOURCES.has(source)) return false;
    if (source !== 'google' || event?.all_day || !interval(event)) return false;
    if (event?.status === 'cancelled' || event?.cancelled === true || event?.transparency === 'transparent') return false;
    return event?.blocking !== false;
  }

  function isDuplicate(first, second) {
    const firstMarker = marker(first);
    const secondMarker = marker(second);
    if (firstMarker && secondMarker && firstMarker === secondMarker) return true;
    const firstId = String(first?.id || '');
    const secondId = String(second?.id || '');
    if (firstId && secondId && firstId === secondId) return true;
    if (!normalizedTitle(first) || normalizedTitle(first) !== normalizedTitle(second)) return false;
    const a = interval(first);
    const b = interval(second);
    if (!a || !b) return false;
    const tolerance = 5 * 60 * 1000;
    return Math.abs(a.start - b.start) <= tolerance && Math.abs(a.end - b.end) <= tolerance;
  }

  function score(event) {
    return [
      isBlocking(event) ? 1 : 0,
      String(event?.source || 'google').toLowerCase() === 'google' ? 1 : 0,
      event?.html_link ? 1 : 0
    ];
  }

  function higherScore(first, second) {
    const a = score(first);
    const b = score(second);
    for (let index = 0; index < a.length; index += 1) {
      if (a[index] !== b[index]) return a[index] > b[index];
    }
    return false;
  }

  function deduplicate(events) {
    const unique = [];
    for (const raw of events || []) {
      const candidate = { ...raw };
      const index = unique.findIndex(existing => isDuplicate(existing, candidate));
      if (index < 0) unique.push(candidate);
      else if (higherScore(candidate, unique[index])) unique[index] = candidate;
    }
    return unique;
  }

  function annotate(events) {
    const cleanEvents = deduplicate(events);
    cleanEvents.forEach(event => {
      event.conflict = false;
      event.conflict_with = [];
      event.blocking = isBlocking(event);
    });
    const candidates = cleanEvents
      .map(event => ({ event, interval: interval(event) }))
      .filter(item => item.event.blocking && item.interval);
    const pairs = [];
    const seen = new Set();

    for (let firstIndex = 0; firstIndex < candidates.length; firstIndex += 1) {
      const first = candidates[firstIndex];
      for (let secondIndex = firstIndex + 1; secondIndex < candidates.length; secondIndex += 1) {
        const second = candidates[secondIndex];
        if (isDuplicate(first.event, second.event)) continue;
        if (first.interval.start < second.interval.end && second.interval.start < first.interval.end) {
          const firstId = String(first.event.id || '');
          const secondId = String(second.event.id || '');
          const key = [firstId, secondId].sort().join(':');
          if (!firstId || !secondId || firstId === secondId || seen.has(key)) continue;
          seen.add(key);
          const overlapMinutes = Math.round((Math.min(first.interval.end, second.interval.end) - Math.max(first.interval.start, second.interval.start)) / 600) / 100;
          pairs.push({ a: firstId, b: secondId, overlapMinutes });
          first.event.conflict = true;
          second.event.conflict = true;
          first.event.conflict_with.push({ id: secondId, title: second.event.title });
          second.event.conflict_with.push({ id: firstId, title: first.event.title });
        }
      }
    }
    return { events: cleanEvents, pairs };
  }

  return { annotate, deduplicate, isBlocking, isDuplicate };
});

File: services/calendar_conflicts.py
import re
from copy import deepcopy
from datetime import datetime, timezone


DERIVED_SOURCES = {'ai', 'deadline', 'email', 'attention', 'work', 'virtual'}


def _parse_datetime(value):
    if not value or len(str(value)) <= 10:
        return None
    try:
        parsed = datetime.fromisoformat(str(value).replace('Z', '+00:00'))
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=timezone.utc)
        return parsed.astimezone(timezone.utc)
    except (TypeError, ValueError):
        return None


def _marker(event):
    harness = event.get('agent_harness') or {}
    return str(
        event.get('marker')
        or event.get('source_message_id')
        or harness.get('agent_harness_marker')
        or harness.get('source_message_id')
        or ''
    ).strip()


def _normalized_title(event):
    return re.sub(r'[^a-z0-9]+', ' ', str(event.get('title') or '')).strip().lower()


def _valid_interval(event):
    start = _parse_datetime(event.get('start'))
    end = _parse_datetime(event.get('end'))
    return (start, end) if start and end and start < end else (None, None)


def is_blocking_event(event):
    source = str(event.get('source') or 'google').strip().lower()
    start, end = _valid_interval(event)
    if source in DERIVED_SOURCES:
        return False
    if source != 'google' or event.get('all_day') or not start or not end:
        return False
    if event.get('status') == 'cancelled' or event.get('cancelled') is True:
        return False
    if event.get('transparency') == 'transparent':
        return False
    return event.get('blocking') is not False


def _same_logical_event(first, second):
    first_marker, second_marker = _marker(first), _marker(second)
    if first_marker and second_marker and first_marker == second_marker:
        return True
    first_id, second_id = str(first.get('id') or ''), str(second.get('id') or '')
    if first_id and second_id and first_id == second_id:
        return True
    if not _normalized_title(first) or _normalized_title(first) != _normalized_title(second):
        return False
    first_start, first_end = _valid_interval(first)
    second_start, second_end = _valid_interval(second)
    if not all((first_start, first_end, second_start, second_end)):
        return False
    tolerance_seconds = 5 * 60
    return (
        abs((first_start - second_start).total_seconds()) <= tolerance_seconds
        and abs((first_end - second_end).total_seconds()) <= tolerance_seconds
    )


def _canonical_score(event):
    return (
        1 if is_blocking_event(event) else 0,
        1 if str(event.get('source') or 'google').lower() == 'google' else 0,
        1 if event.get('html_link') else 0,
    )


def deduplicate_events(events):
    unique = []
    for raw in events or []:
        candidate = deepcopy(raw)
        duplicate_index = next(
            (index for index, existing in enumerate(unique) if _same_logical_event(existing, candidate)),
            None,
        )
        if duplicate_index is None:
            unique.append(candidate)
        elif _canonical_score(candidate) > _canonical_score(unique[duplicate_index]):
            unique[duplicate_index] = candidate
    return unique


def calculate_conflicts(events):
    clean_events = deduplicate_events(events)
    for event in clean_events:
        event['conflict'] = False
        event['conflict_with'] = []
        event['blocking'] = is_blocking_event(event)

    candidates = []
    for event in clean_events:
        start, end = _valid_interval(event)
        if event['blocking'] and start and end:
            candidates.append((event, start, end))

    pairs = []
    seen_pairs = set()
    for index, (first, first_start, first_end) in enumerate(candidates):
        for second, second_start, second_end in candidates[index + 1:]:
            if _same_logical_event(first, second):
                continue
            if first_start < second_end and second_start < first_end:
                first_id, second_id = str(first.get('id') or ''), str(second.get('id') or '')
                pair_key = tuple(sorted((first_id, second_id)))
                if not first_id or not second_id or first_id == second_id or pair_key in seen_pairs:
                    continue
                seen_pairs.add(pair_key)
                overlap = (min(first_end, second_end) - max(first_start, second_start)).total_seconds() / 60
                pairs.append({'a': first_id, 'b': second_id, 'overlapMinutes': round(overlap, 2)})
                first['conflict'] = True
                second['conflict'] = True
                first['conflict_with'].append({'id': second_id, 'title': second.get('title')})
                second['conflict_with'].append({'id': first_id, 'title': first.get('title')})

    return clean_events, pairs
